examples/dpsgd_example.py

```diff
@@ -0,0 +1,185 @@
+   --l2_norm_clip=1.5 \
+   --epochs=15 \
+   --learning_rate=.25 \
+  # this private baseline should get ~96.6% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=1.1 \
+   --l2_norm_clip=1.0 \
+   --epochs=60 \
+   --learning_rate=.15 \
+  # this private baseline should get ~97% acc
+  python -m examples.differentially_private_sgd \
+   --dpsgd=True \
+   --noise_multiplier=0.7 \
+   --l2_norm_clip=1.5 \
+   --epochs=45 \
+   --learning_rate=.25 \
+"""
+
+import time
+import warnings
+
+from absl import app
+from absl import flags
+
+import jax
+import jax.numpy as jnp
+from jax.experimental import stax
+import optax
+
+import datasets
+
+# https://github.com/tensorflow/privacy
+from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp
+from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent
+
+NUM_EXAMPLES = 60_000
+FLAGS = flags.FLAGS
+
+flags.DEFINE_boolean(
+    'dpsgd', True, 'If True, train with DP-SGD. If False, '
+    'train with vanilla SGD.')
+flags.DEFINE_float('learning_rate', .15, 'Learning rate for training')
+flags.DEFINE_float('noise_multiplier', 1.1,
+                   'Ratio of the standard deviation to the clipping norm')
+flags.DEFINE_float('l2_norm_clip', 1.0, 'Clipping norm')
+flags.DEFINE_integer('batch_size', 256, 'Batch size')
+flags.DEFINE_integer('epochs', 60, 'Number of epochs')
+flags.DEFINE_integer('seed', 0, 'Seed for jax PRNG')
+flags.DEFINE_integer(
+    'microbatches', None, 'Number of microbatches '
+    '(must evenly divide batch_size)')
+flags.DEFINE_string('model_dir', None, 'Model directory')
+flags.DEFINE_float('delta', 1e-5, 'Target delta used to compute privacy spent.')
+
+
+init_random_params, predict = stax.serial(
+    stax.Conv(16, (8, 8), padding='SAME', strides=(2, 2)),
+    stax.Relu,
+    stax.MaxPool((2, 2), (1, 1)),
+    stax.Conv(32, (4, 4), padding='VALID', strides=(2, 2)),
+    stax.Relu,
+    stax.MaxPool((2, 2), (1, 1)),
+    stax.Flatten,
+    stax.Dense(32),
+    stax.Relu,
+    stax.Dense(10),
+)
+
+
+def compute_epsilon(steps, target_delta=1e-5):
+  if NUM_EXAMPLES * target_delta > 1.:
+    warnings.warn('Your delta might be too high.')
+  q = FLAGS.batch_size / float(NUM_EXAMPLES)
+  orders = list(jnp.linspace(1.1, 10.9, 99)) + list(range(11, 64))
+  rdp_const = compute_rdp(q, FLAGS.noise_multiplier, steps, orders)
+  eps, _, _ = get_privacy_spent(orders, rdp_const, target_delta=target_delta)
+  return eps
+
+
+def loss_fn(params, batch):
+  logits = predict(params, batch['image'])
+  return optax.softmax_cross_entropy(logits, batch['label']).mean(), logits
+
+
+@jax.jit
+def test_step(params, batch):
+  loss, logits = loss_fn(params, batch)
+  accuracy = (logits.argmax(1) == batch['label'].argmax(1)).mean()
+  return loss, accuracy * 100
+
+
+def main(_):
+  train_dataset = datasets.load_image_dataset('mnist', FLAGS.batch_size)
+  train_dataset = list(train_dataset.as_numpy_iterator())
```

Comment:
When used in the training loop as `train_dataset.as_numpy_iterator()`, the time per private epoch is ~4 seconds, which is significantly slower than this using list (which results in a per private epoch time of ~0.7 seconds). I tried using `prefetch`, which didn't help. How else can I improve this?